include/envoy/http/filter.h

```diff
@@ -190,6 +190,11 @@ class StreamDecoderFilterCallbacks : public virtual StreamFilterCallbacks {
    */
   virtual void addDecodedData(Buffer::Instance& data, bool streaming_filter) PURE;
 
+  /**
+   * Adds decoded trailers. This will overwrite any existing trailers should any already exist.
+   */
+  virtual void addDecodedTrailers(HeaderMapPtr&& trailers) PURE;
```

Comment:
still -1 to moving a pointer here, and prefer the first revision: `HeaderMap& addDecodedTrailers()` for reasons:

- moving a pointer doesn't prevent filters from keeping reference to `HeaderMap` itself too long, so that still can happen. Returning a reference is already clear that Envoy core (ActiveStream) owns the `HeaderMap`.
- the filter will now have to know implementations of HeaderMap (i.e. HeaderMapImpl) and instantiate, I know there is already a couple places doing this, mostly to deal with `AsyncHttpClient`, perhaps we shouldn't add another place requiring that? 

@mattklein123 thoughts?

Reply:
> moving a pointer doesn't prevent filters from keeping reference to HeaderMap itself too long, so that still can happen. Returning a reference is already clear that Envoy core (ActiveStream) owns the HeaderMap.

IMO move semantics make it clear that the pointer is now owned by the callee. We do this in a variety of places.

> the filter will now have to know implementations of HeaderMap (i.e. HeaderMapImpl) and instantiate, I know there is already a couple places doing this, mostly to deal with AsyncHttpClient, perhaps we shouldn't add another place requiring that?

I think this is a better argument. Assuming with the new semantics we have discussed it's just as easy to return a reference to either the existing trailers or a newly created trailers, it sounds reasonable to me to switch it back to what we had before.

Reply:
The one thing that's unclear to me is what `HeaderPtr& addEncodedTrailers` would return if called before `encodeTrailers(..., true)`? would we just throw/assert?

Reply:
Yes, I would throw some type of bad argument exception that will cause Envoy to crash and can be checked in tests. (I would probably not derive from EnvoyException and pick a std exception for this).